### PR TITLE
Mpi & mpi communicator

### PR DESCRIPTION
**Pull request by amgrigoriev**

### CODE REVIEW

File: cpp/oneapi/dal/detail/mpi/communicator.hpp [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+
+        ONEDAL_ASSERT(send_buf);
+        if (get_rank() == root) {
+            ONEDAL_ASSERT(recv_counts);
+            ONEDAL_ASSERT(displs);
+            ONEDAL_ASSERT(recv_buf);
+        }
+
+        array<int> recv_counts_int;
+        array<int> displs_int;
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/detail/mpi/communicator.hpp:199]: Please don't use `int`. Can you use `std::int32_t` instead?

**Review by KulikovNikita — APPROVED** [review]

File: cpp/oneapi/dal/detail/communicator.cpp [hunk 2]
```diff
@@ -18,10 +18,283 @@
...
+        if (alloc_kind == sycl::usm::alloc::unknown) {
+            throw invalid_argument{ error_messages::unknown_usm_pointer_type() };
+        }
+    }
+}
+#endif
+
+#ifdef ONEDAL_DATA_PARALLEL
+static void wait_request(spmd_request_iface* request) {
+    if (request) {
```
[the remark this answers is not in the document]
Comment (makart19) [reply cpp/oneapi/dal/detail/communicator.cpp:34]: then it should be 
if (request != nullptr)

File: cpp/oneapi/dal/detail/mpi/communicator.hpp [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+        default: ONEDAL_ASSERT(!"Unknown data type");
+    }
+    return MPI_DATATYPE_NULL;
+}
+
+inline MPI_Op make_mpi_reduce_op(const spmd_reduce_op& op) {
+    switch (op) {
+        case spmd_reduce_op::sum: return MPI_SUM;
+        default: ONEDAL_ASSERT(!"Unknown reduce operation");
+    }
```
[the remark this answers is not in the document]
Comment (amgrigoriev) [reply cpp/oneapi/dal/detail/mpi/communicator.hpp:55]: A list of "op" is expected

File: cpp/oneapi/dal/detail/mpi/communicator.hpp (hunk 1 above)
Comment (amgrigoriev) [reply cpp/oneapi/dal/detail/mpi/communicator.hpp:199]: Should be int for MPI

File: cpp/oneapi/dal/detail/communicator.cpp [hunk 4]
```diff
@@ -18,10 +18,283 @@
...
+    wait_request(gatherv(send_buff_host.get_data(),
+                         send_count,
+                         recv_buf_host_ptr,
+                         recv_counts_host,
+                         displs_host_0.get_data(),
+                         dtype,
+                         root));
+
+    if (get_rank() == root) {
+        const std::int64_t* displs_host_0_ptr = displs_host_0.get_data();
```
[the remark this answers is not in the document]
Comment (amgrigoriev) [reply cpp/oneapi/dal/detail/communicator.cpp:202]: should be "_root"

File: cpp/oneapi/dal/detail/communicator.cpp [hunk 5]
```diff
@@ -18,10 +18,283 @@
...
+#ifdef ONEDAL_DATA_PARALLEL
+static void check_if_pointer_matches_queue(const sycl::queue& q, const void* ptr) {
+    if (ptr) {
+        const auto alloc_kind = sycl::get_pointer_type(ptr, q.get_context());
+        if (alloc_kind == sycl::usm::alloc::unknown) {
+            throw invalid_argument{ error_messages::unknown_usm_pointer_type() };
+        }
+    }
+}
+#endif
```
[the remark this answers is not in the document]
Comment (amgrigoriev) [reply cpp/oneapi/dal/detail/communicator.cpp:30]: "is_known_usm"

[the remark this answers is not in the document]
Comment (amgrigoriev) [reply cpp/oneapi/dal/detail/communicator.cpp:30]: But the method is defined in backend. So it make sense for now to remain the code "as is"

File: cpp/oneapi/dal/detail/communicator.cpp [hunk 6]
```diff
@@ -18,10 +18,283 @@
...
+
+#ifdef ONEDAL_DATA_PARALLEL
+auto spmd_communicator_via_host_impl::allgather(sycl::queue& q,
+                                                const byte_t* send_buf,
+                                                std::int64_t send_count,
+                                                byte_t* recv_buf,
+                                                std::int64_t recv_count,
+                                                const data_type& dtype,
+                                                const std::vector<sycl::event>& deps)
+    -> spmd_request_iface* {
```
[the remark this answers is not in the document]
Comment (amgrigoriev) [reply cpp/oneapi/dal/detail/communicator.cpp:228]: it is wrapped in pimpl on next level

**Review by makart19 — APPROVED** [review]

**Review by KulikovNikita — APPROVED** [review]

**Pull request merged**